### PR TITLE
feat(auth): Return ID token from /token endpoint

### PR DESCRIPTION
**Pull request by EricHasegawa**
Returns an HS256 encrypted ID token from the `/oauth/token` endpoint if the access token has the right scopes. 
Also:
- Adds the necessary scopes to achieve this.
- Refactors the `token` code to avoid having one really long `POST` method.
- Improves testing for the `token` endpoint.
- Uses a secure random value to encrypt ID tokens until we can implement proper shared secrets (i.e. this is safe to release).

### CODE REVIEW

File: src/sentry/web/frontend/oauth_token.py
Comment (markstory) [inline src/sentry/web/frontend/oauth_token.py:88]:
```suggestion
                grant.user_id,
```

Accessing `.user` will emit another SQL query.

File: src/sentry/web/frontend/oauth_token.py
Comment (markstory) [inline src/sentry/web/frontend/oauth_token.py:117]:
```suggestion
            return self.error(request, "invalid_token", "invalid refresh token")
```

This error message should reflect the problem.

File: src/sentry/web/frontend/oauth_token.py
Comment (markstory) [inline src/sentry/web/frontend/oauth_token.py:93]: Won't the ID token also need to know which scopes are authorized so that email/profile scope details can be included as required?

Comment (EricHasegawa) [reply src/sentry/web/frontend/oauth_token.py:93]: The ID token actually just exists on its own mostly, its not used to access API's or protected resources at any point. It's meant to be read by the client upon a users authentication, and it just contains information about what happened during a users authentication + confirmation that the authentication was successful ([source](https://oauth.net/id-tokens-vs-access-tokens/#:~:text=An%20ID%20token%20contains%20information,requirement%20of%20an%20ID%20token.)). The fact that we put email in here is just a "nice to have", it's not required in the spec.

The followup PR to this one ([here](https://github.com/getsentry/sentry/pull/52493)) adds the `/userinfo` endpoint which is the other major part of OIDC, that's a specialized protected resource where clients with the `oidc` scope can retrieve more details on the user, and that's accessed with the standard access token.

File: src/sentry/web/frontend/oauth_token.py
Comment (EricHasegawa) [reply src/sentry/web/frontend/oauth_token.py]: Good catch!

File: src/sentry/web/frontend/oauth_token.py
Comment (markstory) [reply src/sentry/web/frontend/oauth_token.py:93]: Yeah, ID tokens aren't used to access APIs/protected resources, but they can also contain profile/email data ([example of that](https://auth0.com/blog/id-token-access-token-what-is-the-difference/)). If profile/email data is part of the ID token, client applications can avoid using `/userinfo` most of the time.

Comment (EricHasegawa) [reply src/sentry/web/frontend/oauth_token.py:93]: Yeah good point - Since it's a separable piece of work from this one I implemented it in a new PR [here](https://github.com/getsentry/sentry/pull/52556). If you don't want this PR to go through without that functionality we can merge the ID token details PR into this one first, or we can merge this one first then do that one after - your call!

Comment (markstory) [reply src/sentry/web/frontend/oauth_token.py:93]: > If you don't want this PR to go through without that functionality we can merge the ID token details PR into this one first, or we can merge this one first then do that one after - your call!

I'm not too fussy. Great to see that our ID tokens will have the user profile data in it so that consumers aren't required to call `/userinfo` as well.

File: tests/sentry/web/frontend/test_oauth_token.py
Comment (markstory) [inline tests/sentry/web/frontend/test_oauth_token.py:199]: Should this test do assertions on the redirect uri? The test name implies that would  happen.

**Review by markstory — APPROVED** [review]

File: tests/sentry/web/frontend/test_oauth_token.py
Comment (EricHasegawa) [reply tests/sentry/web/frontend/test_oauth_token.py:199]: There's no redirect URI or anything returned here - basically we're checking that whatever redirect URI was passed to `/authorize` to generate the authorization code is also passed to `/token`. So if we don't pass a redirect URI to `/authorize` we use the default, and therefore we don't need to pass one to `/token` since it will check against the default URI. This test just makes sure it doesn't error out in that case.

**Pull request merged**
